workspaces/ballerina/ballerina-extension/src/features/ai/service/code/code.ts

```diff
@@ -16,48 +16,75 @@
 
 import { CoreMessage, generateText, streamText } from "ai";
 import { getAnthropicClient, ANTHROPIC_SONNET_4, getProviderCacheControl } from "../connection";
-import { GenerationType, getRelevantLibrariesAndFunctions } from "../libs/libs";
-import { getRewrittenPrompt, populateHistory, transformProjectSource, getErrorMessage, extractResourceDocumentContent } from "../utils";
-import { getMaximizedSelectedLibs, selectRequiredFunctions, toMaximizedLibrariesFromLibJson } from "./../libs/funcs";
-import { GetFunctionResponse } from "./../libs/funcs_inter_types";
+import { GenerationType, getAllLibraries } from "../libs/libs";
+import { getLibraryProviderTool } from "../libs/libraryProviderTool";
+import {
+    getRewrittenPrompt,
+    populateHistory,
+    transformProjectSource,
+    getErrorMessage,
+    extractResourceDocumentContent,
+} from "../utils";
 import { LANGLIBS } from "./../libs/langlibs";
 import { Library } from "./../libs/libs_types";
 import {
-    ChatNotify,
     DiagnosticEntry,
     FileAttatchment,
     GenerateCodeRequest,
-    onChatNotify,
     OperationType,
     PostProcessResponse,
-    ProjectDiagnostics,
     ProjectSource,
     RepairParams,
     RepairResponse,
     SourceFiles,
-    Command
+    Command,
 } from "@wso2/ballerina-core";
 import { getProjectSource, postProcess } from "../../../../rpc-managers/ai-panel/rpc-manager";
 import { CopilotEventHandler, createWebviewEventHandler } from "../event";
 import { AIPanelAbortController } from "../../../../../src/rpc-managers/ai-panel/utils";
 import { getRequirementAnalysisCodeGenPrefix, getRequirementAnalysisTestGenPrefix } from "./np_prompts";
 
+
+function appendFinalMessages(history: CoreMessage[], finalMessages: CoreMessage[]): void {
+    const knownIds = new Set<string>();
+    for (let i = 0; i < finalMessages.length - 1; i++) {
+        const message = finalMessages[i];
+        const messageId = (message as any).id;
+
+        if ((message.role === "assistant" || message.role === "tool") && messageId && !knownIds.has(messageId)) {
+            knownIds.add(messageId);
+
+            if (i === finalMessages.length - 2) {
+                message.providerOptions = {
```

Comment:
We need to consider bedrock caching case as well, shall we use the function instead?